#' Load gene-model annotations compiled from GTF format
#'
#' @description Load gene-model annotations for two-way plotting in per_sample and domain tab-panels
#'
#' @param ensemblId A string - ensembl gene id (e.g. \emph{ENSG00000141510}).
#' @param txdb_ref An object - an annotation database engine loaded from *.sqlite file (initially created by compiling a GTF annotation file).
#' @param grTrack_ref An object - annotation tracks loaded from *.Rd file (originally generated by processing \code{txdb_ref} object).
#'
#' @return A list (e.g. object_A and object_B) to collect transcript annotations of partner genes
#'
#' @export
get_annotation_db <- function(ensemblId, txdb_ref, grTrack_ref) {
	# For testing: ensemblId = ensemblId_A; txdb_ref = txdb;
	chrom_only = c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22","chrX","chrY","chrM");
	cols = c("TXNAME", "TXCHROM", "TXSTRAND", "TXSTART", "TXEND");
	dataset = AnnotationDbi::select(txdb_ref, keys=ensemblId, columns=cols, keytype="GENEID");

	#// get the start and end coordinate of partner genes
	start = tapply(dataset$TXSTART, as.factor(dataset$GENEID), min);
	end = tapply(dataset$TXEND, as.factor(dataset$GENEID), max);
	strand = tapply(dataset$TXSTRAND, as.factor(dataset$GENEID), max);
	chr = tapply(dataset$TXCHROM, as.factor(dataset$GENEID), max);

	#// judge the status of 'chr', if not chromosome, return NULL
	if (! chr %in% chrom_only ) { return(NULL); } 
	#// get the annotations and structures (e.g. exon info) of merged transcripts for gene partner
	#-# chr_ucsc= paste("chr", chr, sep="") # use chrX instead of X here
	subTrack <- Gviz::subset(grTrack_ref[[chr]], from=start, to=end);

	#// subTrack is a GRange class, and dataset is a data.frame class
	tmp <- list(txTr_f=subTrack, dataset=dataset, chr=chr, start=start, end=end, strand=strand);
	return(tmp);
}
	
#' Load gene-model annotations compiled from GTF format
#'
#' @description Load gene-model annotations for two-way plotting in overview tab-panel
#'
#' @param ensemblId A string - ensembl gene id (e.g. \emph{ENSG00000141510}).
#' @param txdb_ref An object - annotation database engine loaded from *.sqlite file (initially created by compiling a GTF annotation file).
#' @param grTrack_ref An object - annotation tracks loaded from *.Rd file (originally generated by processing \code{txdb_ref} object).
#' @param whole_tx A GRange object - exon interval grouped by transcript per gene, which was generated by processing \code{txdb_ref} object.
#'
#' @return A list (e.g. object_A and object_B) to collect transcript annotations of partner genes
#'
#' @export
get_annotation_db_extend <- function(ensemblId, txdb_ref, grTrack_ref, whole_tx) {
	# For testing: ensemblId = ensemblId_B; txdb_ref = txdb; grTrack_ref = grTrack; whole_tx = whole_txdb
	chrom_only = c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22","chrX","chrY","chrM");
	cols = c("TXNAME", "TXCHROM", "TXSTRAND", "TXSTART", "TXEND");
	dataset = AnnotationDbi::select(txdb_ref, keys=ensemblId, columns=cols, keytype="GENEID");

	#// get the start and end coordinate of partner genes
	start = tapply(dataset$TXSTART, as.factor(dataset$GENEID), min);
	end = tapply(dataset$TXEND, as.factor(dataset$GENEID), max);
	strand = tapply(dataset$TXSTRAND, as.factor(dataset$GENEID), max);
	chr = tapply(dataset$TXCHROM, as.factor(dataset$GENEID), max);

	#// judge the status of 'chr', if not chromosome, return
	if (! chr %in% chrom_only ) { return(NULL); }
	#// get the annotations and structures (e.g. exon info) of merged transcripts for gene partner
	#-# chr_ucsc= paste("chr", chr, sep="") # use chrX instead of X here
	subTrack <- Gviz::subset(grTrack_ref[[chr]], from=start, to=end);

	#// assign the correct id_name to each exon of annotated transcripts (structured as "5-utr-cds-utr-3")
	select_region_f = c(); 
	for ( name in dataset$TXNAME ) {
		if ( strand == "+" ) {
			all_exon_tx_A = whole_tx[[name]]
		} else { #// if negative strand, reverse exon order
			all_exon_tx_A = rev(whole_tx[[name]])
		}
		#// run GRange overlapping
		ref_point = GenomicRanges::findOverlaps(subTrack[Gviz::transcript(subTrack)==name]@range, all_exon_tx_A)
		#// set exon_name 
		subTrack@range@elementMetadata[subTrack@range@elementMetadata$transcript==name,]$exon[S4Vectors::queryHits(ref_point)] = all_exon_tx_A$exon_name[S4Vectors::subjectHits(ref_point)]
		#// convert GeneRegionTrack(Rle) to data.frame structure
		region_f = as.data.frame(subTrack[Gviz::transcript(subTrack)==name]@range)
		select_region_f = rbind(select_region_f, region_f)
	}
	#// rank transcripts from the longest to the shortest
	#  dataset = dplyr::arrange(dataset, desc(TXEND-TXSTART)) 
	dataset = dataset[order(-(dataset$TXEND-dataset$TXSTART)), ];

	#// select_region_f is a data.frame class, and dataset is a data.frame class
	tmp <- list(txTr_f=select_region_f, dataset=dataset, chr=chr, start=start, end=end, strand=strand)
	return(tmp)
}

